docs/architecture.md

```diff
@@ -0,0 +1,119 @@
+Architecture
+===
+
+Fixed-structure heterogeneous entities with no possibility of adding or removing
+components on runtime.
+
+# Entities
+Entities are structures that share a common header (preamble), as following:
+
+    Entity
+      |
+      +-- type_id: int
+      +-- flags: int
+      +-- x: int
+      +-- y: int
+      +-- userdata: ptr
+
+An example of a derived entity:
+
+    Player : Entity
+      |   ; Common fields
+      +-- type_id: int
+      +-- flags: int
+      +-- x: int
+      +-- y: int
+      +-- userdata: ptr to any
+      |   ; Player fields
+      +-- movement: MovementComponent
+      +-- health: HealthComponent
+      +-- ship: SpriteComponent
+      +-- flame: SpriteComponent
+
+
+# Components
+Components are arbitrary structures, the only thing in common they share is a
+pointer to the parent entity:
+
+    Component:
+      |
+      +-- entity: ptr to Entity
+
+
+An example component:
+
+    HealthComponent : Component
+      |   ; Component fields
+      +-- entity: ptr to Entity
+      |   ; HealthComponent fields
+      +-- hp: int
+
+
+# Entity ops
+Entity logic is defined by a set of associated functions (ops), which map to the
+value of `type_id` field.
+
+    EntityOps:
+      |
+      +-- init()  ; called on entity creation
+      +-- fini()  ; called upon entity destruction
+      +-- tick()  ; called each frame
+
```

Comment:
Atm there are two ways to create an entity
-  loading its data from the rom to the ram 
- dynamic creation of entity
the init op is used to initialize variables if needed, to register components in lists etc.